File: server/plugins/image_group.py
"""图片组节点 - 管理底图、素材图和多张参考图。"""

import json


NODE_DEF = {
    "def_id": "image-group",
    "name": "图片组",
    "category": "素材",
    "view": "image-group",
    "inputs": [],
    "outputs": [
        {"id": "image", "label": "当前图片", "type": "IMAGE"},
        {"id": "background", "label": "底图", "type": "IMAGE"},
        {"id": "foreground", "label": "素材", "type": "IMAGE"},
        {"id": "references", "label": "参考图", "type": "IMAGE_LIST"},
        {"id": "images", "label": "全部图片", "type": "IMAGE_LIST"},
    ],
    "controls": [
        {"kind": "text", "id": "items", "label": "图片", "default": "[]"},
        {"kind": "text", "id": "selected_id", "label": "当前图片", "default": ""},
    ],
}


def _parse_items(value) -> list[dict]:
    if isinstance(value, str):
        try:
            value = json.loads(value)
        except json.JSONDecodeError:
            return []
    if not isinstance(value, list):
        return []
    return [
        item for item in value
        if isinstance(item, dict) and isinstance(item.get("url"), str) and item["url"]
    ]


async def process(inputs: dict, controls: dict, context: dict) -> dict:
    items = _parse_items(controls.get("items"))
    selected_id = controls.get("selected_id")
    selected = next((item for item in items if item.get("id") == selected_id), None)
    if selected is None and items:
        selected = items[0]

    return {
        "image": selected.get("url") if selected else None,
        "background": next((item["url"] for item in items if item.get("role") == "background"), None),
        "foreground": next((item["url"] for item in items if item.get("role") == "foreground"), None),
        "references": [item["url"] for item in items if item.get("role") == "reference"],
        "images": [item["url"] for item in items],
    }

File: src/nodes/imageGroup.ts
import type { PortValues } from '../types/workflow';

export type ImageGroupRole = 'background' | 'foreground' | 'reference';

export type ImageGroupItem = {
  id: string;
  url: string;
  role: ImageGroupRole;
  name?: string;
};

const VALID_ROLES = new Set<ImageGroupRole>(['background', 'foreground', 'reference']);

export const IMAGE_GROUP_ROLE_LABELS: Record<ImageGroupRole, string> = {
  background: '底图',
  foreground: '素材',
  reference: '参考',
};

export function parseImageGroupItems(value: unknown): ImageGroupItem[] {
  if (!value) return [];
  try {
    const parsed = typeof value === 'string' ? JSON.parse(value) : value;
    if (!Array.isArray(parsed)) return [];
    return parsed.flatMap((item): ImageGroupItem[] => {
      if (!item || typeof item !== 'object') return [];
      const candidate = item as Partial<ImageGroupItem>;
      if (typeof candidate.url !== 'string' || !candidate.url) return [];
      return [{
        id: typeof candidate.id === 'string' && candidate.id ? candidate.id : crypto.randomUUID(),
        url: candidate.url,
        role: VALID_ROLES.has(candidate.role as ImageGroupRole) ? candidate.role as ImageGroupRole : 'reference',
        name: typeof candidate.name === 'string' ? candidate.name : undefined,
      }];
    });
  } catch {
    return [];
  }
}

export function normalizeImageGroupItems(items: ImageGroupItem[]): ImageGroupItem[] {
  let hasBackground = false;
  let hasForeground = false;
  return items.map((item) => {
    if (item.role === 'background') {
      if (hasBackground) return { ...item, role: 'reference' };
      hasBackground = true;
    }
    if (item.role === 'foreground') {
      if (hasForeground) return { ...item, role: 'reference' };
      hasForeground = true;
    }
    return item;
  });
}

export function createImageGroupOutputs(items: ImageGroupItem[], selectedId?: string | null): PortValues {
  const selected = items.find((item) => item.id === selectedId) ?? items[0];
  return {
    'output-image': selected?.url ?? null,
    'output-background': items.find((item) => item.role === 'background')?.url ?? null,
    'output-foreground': items.find((item) => item.role === 'foreground')?.url ?? null,
    'output-references': items.filter((item) => item.role === 'reference').map((item) => item.url),
    'output-images': items.map((item) => item.url),
  };
}
